### PR TITLE
[numeral] [plugins] Remove Coq's `Bigint` module in favor of ZArith.

### PR DESCRIPTION
**Pull request by ejgallego**
We replace Coq's use of `Bigint`, `Big_int` and `num` by the ZArith OCaml
library which is a more modern version.

We switch the core files and easy plugins only for now, the remaining `micromega` plugins will be done in their own commit.

We thus keep the num library linked for now until micromega is ported.

Co-authored-by: Vincent Laporte <Vincent.Laporte@fondation-inria.fr>

**Kind:** feature / infrastructure.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated
- [x] Entry added in the changelog

Overlays:
- https://github.com/coq/bignums/pull/44

### CODE REVIEW

**Comment by ejgallego** [comment]
This PR ports Coq to ZArith except for micromega; it is the first part of #8743 ;

The porting is actually quite ortogonal so we could indeed make #8743 self-contained for micromega, but IMO I think this is close enough to be ready that it is not worth it.

**Comment by ejgallego** [comment]
Two problems:

- windows build needs to install gmp + zarith
- coq_makefile using -linkall makes linking fail in OCaml >= 4.08, workaround pushed.

Bench : https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/875/

**Comment by ejgallego** [comment]
@maximedenes can we install libgmp-dev on pendulum? [I'm afraid my network situation is bad as I'm on the road]

**Comment by maximedenes** [comment]
> @maximedenes can we install libgmp-dev on pendulum? [I'm afraid my network situation is bad as I'm on the road]

Done.

**Comment by maximedenes** [comment]
I have a fear releated to this PR: since the porting of micromega failed, how do we know that this is correct? Aren't we just porting less well-tested parts (`Big_int` is notorious for having has some undetected long-standing critical bugs, for example, IIRC)? I would feel more comfortable if we port everything at once.

Also, what is the motivation for this change if `num` is still a dependency?

**Comment by ejgallego** [comment]
> I have a fear related to this PR: since the porting of micromega failed how do we know that this is correct?

I'm not sure I understand what you mean about correctness; what is the particular concern you have in mind?

I don't think porting of micromega "failed", it just needs more work as there is not a 1-1 mapping of the APIs in a few places.

> Aren't we just porting less well-tested parts (`Big_int` is notorious for having has some undetected long-standing critical bugs, for example, IIRC)?

What does `Big_int` has to do here? It is not involved in this PR IIANM; what do you mean about less "well-tested" parts? Both omega and the numeral notations are pretty well tested, and I'd say zarith/GMP is way more tested than num/bigint. 

> I would feel more comfortable if we port everything at once.

These are totally orthogonal components so I don't see any reason porting more stuff should increase the confidence on the current code.

**Comment by maximedenes** [comment]
@ejgallego my worry is not about the correctness of ZArith/GMP, of course, it is precisely about the correctness of the API mapping.

> What does `Big_int` has to do here? It is not involved in this PR

I imagined you were replacing the uses of Big_int by ZArith because that's literally what the PR title says, but I must have misunderstood.

Under that assumption, my point was that uses of `Big_int` were not well tested. So the replacement by `ZArith` will also not be well tested.

**Comment by ejgallego** [comment]
> I imagined you were replacing the uses of Big_int by ZArith because that's literally what the PR title says, but I must have misunderstood.

This is replacing `Bigint`, which is Coq's own version, `Big_int` is the one in OCaml [confusing I know]

> @ejgallego my worry is not about the correctness of ZArith/GMP, of course, it is precisely about the correctness of the API mapping.
>
> Under that assumption, my point was that uses of `Big_int` were not well tested. So the replacement by `ZArith` will also not be well tested.

Well that's hard to know, tho the API mapping performed this PR seems safe enough to me; I don't think the paths this PR modifies are not so well tested tho, they include omega and numeral notations, these are certainly used in quite a few places in the CI.

**Comment by ejgallego** [comment]
> Done.

Thanks, let's see the new bench here https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/876/

**Comment by ejgallego** [comment]
By the way micromega mostly works, I just don't have the time to fix it now, but it should not be hard. The problem is almost for sure in the particular spot where exponentation of rationals was used, but indeed there the old num API was allowing clients to do what seems like very weird things; so indeed we will have to improve code likely.

Note also that micromega uses `big_int` vs the parts of Coq that don't use it.

**Comment by SkySkimmer** [comment]
Looks like it needs an overlay for bignums, and there's an anomaly in the test suite.

**Review by Zimmi48 — APPROVED** [review]
I'm not very fond of the intermediate state where we depend on both `num` and `ZArith`, I seriously hope that we won't release a version with both of these dependencies. Yet, I won't stand in the way of this PR.

File: dev/ci/docker/bionic_coq/Dockerfile
Comment (Zimmi48) [inline dev/ci/docker/bionic_coq/Dockerfile]:
```suggestion
# Common OPAM packages, num to be removed once the migration to zarith is complete
```

**Comment by ejgallego** [comment]
> I'm not very fond of the intermediate state where we depend on both `num` and `ZArith`, I seriously hope that we won't release a version with both of these dependencies. Yet, I won't stand in the way of this PR.

Actually it makes sense to do the migration in gradual steps, as the uses of num in the codebase are really independent. But indeed I agree we should maybe coordinate the PRs so no release depends on both.

**Comment by ejgallego** [comment]
What should we do with this? Maybe getting an assignee would be the first step?

**Comment by silene** [comment]
What is the status of jsCoq nowadays? Is it still supposed to work correctly after this pull request?

**Comment by ejgallego** [comment]
> What is the status of jsCoq nowadays? Is it still supposed to work correctly after this pull request?

I understand there is an implementation of `zarith` for js_of_ocaml

**Comment by silene** [comment]
To be clear, big integers were added to the [JavaScript standard](https://tc39.es/ecma262/) only two days ago. So, it is still a moving target, as far as implementations are concerned.

**Comment by ejgallego** [comment]
> To be clear, big integers were added to the [JavaScript standard](https://tc39.es/ecma262/) only two days ago. So, it is still a moving target, as far as implementations are concerned.

The zarith implementation was not using native js big integers but some other library.

**Review by maximedenes — COMMENTED** [review]
How about turning this PR into a draft until it is ready for review?

**Comment by ejgallego** [comment]
Actually this PR is mostly ready in terms of code [modulo bugs], what is not clear to me is how do people want to proceed with the zarith transition, if they want to proceed at all with it.

There are 3 choices:
- don't do anything, keep using [num] , possibly vendoring it
- merge this, and live with both num and zarith until the micromega part is ready. IMHO not a big deal, as the uses are completely orthogonal
- wait until micromega is ported to merge this.

Note that the use of `num/zarith` in micromega is completely independent from the use in the parts touched in this PR.

**Comment by maximedenes** [comment]
I'd vote for the second option (merge this, and live with both num and zarith until the micromega part is ready).

**Comment by ejgallego** [comment]
Ok, I will then update this PR and evaluate its status.

**Comment by maximedenes** [comment]
Thanks, I'll assign myself then, if that's ok.

**Comment by ejgallego** [comment]
> Thanks, I'll assign myself then, if that's ok.

Sounds good!

**Comment by ejgallego** [comment]
@maximedenes so this seems to mostly work, IMHO we can move to review while I fix windows and add the changelog.

Nix also needs an update.

**Comment by ejgallego** [comment]
Down to 3 failures:

- windows azure: No idea what's going on, it seems I am not setting some cygwin magic correctly, `conf-gmp` doesn't detect `gmp`.
- windows gitlab: No idea what's going on, it is impossible for me to debug that script I'm afraid.
- nix, also no idea what is going on.

I can take care of the first one, but definitively need help with the other two.

File: dev/build/windows/makecoq_mingw.sh
Comment (MSoegtropIMC) [inline dev/build/windows/makecoq_mingw.sh:1030]: @ejgallego : this does not work. log1 assumes that the first argument is at the same time a command and a valid *piece* of a log file name. If the command contains e.g. /, you should use logn, where the first argument is an arbitrary log name and the rest is the command. Typical usage is `logn configure ./configure`.

This is easy to see in the log - just scroll above the artefact upload (which looks very regular).

**Comment by MSoegtropIMC** [comment]
> * windows gitlab: No idea what's going on, it is impossible for me to debug that script I'm afraid.

@ejgallego : I left a comment what is most likely wrong.

File: dev/build/windows/makecoq_mingw.sh
Comment (ejgallego) [reply dev/build/windows/makecoq_mingw.sh:1030]: Thanks, I was indeed confused by this, if I tried `log1 configure` it said "configure program not found".

File: Makefile.build
Comment (vbgl) [inline Makefile.build:763]: Why is this part removed? `micromega` still depends on `num`.

Comment (ejgallego) [reply Makefile.build:763]: Indeed, it seems a better idea to keep it for now.

File: dev/build/windows/makecoq_mingw.sh
Comment (MSoegtropIMC) [reply dev/build/windows/makecoq_mingw.sh]: > Thanks, I was indeed confused by this, if I tried `log1 configure` it said "configure program not found".

yes, as you know this is usual posix shell behaviour.

In all non trivial cases use logn (for log name). log1 and log2 are just abbreviations which reuse the first (and second) word of the command as log file name postfix(es). logn is the generic variant.

**Comment by MSoegtropIMC** [comment]
> Well that's hard to know, tho the API mapping performed this PR seems safe enough to me; I don't think the paths this PR modifies are not so well tested tho, they include omega and numeral notations, these are certainly used in quite a few places in the CI.

I have some fairly hard tests for Coq's machine integer arithmetic based on some high precision computations in CoqInterval (stuff like comparing different ways to compute pi to 1000 bits or a correctness proof for a 140 bit polynomial approximation of sin for x in 0...pi/2). I don't know if this is eventually based on Bigint or not. If it is, I could do a quick PR to add these tests to the test suite of CoqInterval (I should anyway do this, because there were two proofs of False in the past cause of wrong arithmetic).

**Comment by proux01** [comment]
> I don't know if this is eventually based on Bigint or not.

For parsing and printing yes, not for actual computations (BigZ being implemented on top of primitive integers Int63). So I'm not expecting much performance difference.

**Comment by ejgallego** [comment]
Down to 1 failure:

> * windows azure: No idea what's going on, it seems I am not setting some cygwin magic correctly, `conf-gmp` doesn't detect `gmp`.

```
#=== ERROR while compiling conf-gmp.1 =========================================#
# context              2.0.7 | win32/x86_64 | ocaml-variants.4.10.0+mingw64c | git+https://github.com/fdopen/opam-repository-mingw.git#opam2
# path                 ~/.opam/ocaml-variants.4.10.0+mingw64c/.opam-switch/build/conf-gmp.1
# command              ./test-win.sh
# exit-code            1
# env-file             ~/.opam/log/conf-gmp-2800-fbcac7.env
# output-file          ~/.opam/log/conf-gmp-2800-fbcac7.out
### output ###
# test.c:1:10: fatal error: gmp.h: No such file or directory
#     1 | #include <gmp.h>
#       |          ^~~~~~~
# compilation terminated.
```
still not clear why this is failing, but firing up my VM should solve it.

**Comment by MSoegtropIMC** [comment]
> still not clear why this is failing, but firing up my VM should allow to debug it.

@ejgallego : I would guess you are missing these option when installing cygwin:

```
    -P mingw64-%ARCH%-gmp,mingw64-%ARCH%-mpfr ^
```

You might want to have a look at (https://github.com/coq/coq/blob/master/dev/build/windows/MakeCoq_MinGW.bat) to see what packages I add when installing cygwin.

Afaik there are no *official* ways to install packages from a shell after the facts - you need to do this when installing cygwin. But there are some unofficial tools which can do this - including one for opam. I would have to look this up in case you can't easily add something to the cygwin install.

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml:108]:
```suggestion
  let string_of_nonneg_bigint c n =
    match c with
    | CDec -> Z.format "%d" n
    | CHex -> "0x" ^ Z.format "%x" n
```
(and delete all remaining of the function)

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml:91]:
```suggestion
    Z.of_string (to_string (sign,n))
```
(and remove all the remaining of the function)

**Review by proux01 — COMMENTED** [review]
Just two simplification since Zarith handles hexadecimals.

File: interp/numTok.ml
Comment (ejgallego) [reply interp/numTok.ml:108]: Done, thanks!

**Comment by ejgallego** [comment]
> -P mingw64-%ARCH%-gmp,mingw64-%ARCH%-mpfr

That did it, thanks @MSoegtropIMC !

**Comment by ejgallego** [comment]
Ok, this should be ready for review and eventual merge, thanks to all!

**Comment by liyishuai** [comment]
Is `big.ml` tested anywhere? I was proposing to remove `Big` in #8252.

**Review by liyishuai — CHANGES_REQUESTED** [review]
```coq
Require Import ExtrOcamlZBigInt ZArith.
Fail Extraction TestCompile Z_modulo_2.
```
```ocaml
File "/var/folders/lm/cpf87_lx21n9bgnl4kr72rjm0000gn/T/testextraction1c373d.mli", line 13, characters 13-24:
13 |   val succ : Big.big_int -> Big.big_int
                  ^^^^^^^^^^^
Error: Unbound module Big
```
> The command has indeed failed with message:
Compilation of file /var/folders/lm/cpf87_lx21n9bgnl4kr72rjm0000gn/T/testextraction1c373d.ml failed with exit code 2

**Comment by ejgallego** [comment]
> Is `big.ml` tested anywhere? I was proposing to remove `Big` in #8252.

I'm not sure this PR does modify anything regarding `plugins/extraction/big.ml` ; but indeed should maybe make your PR much easier to merge, as now `zarith` will be in scope IIUC.

**Review by liyishuai — APPROVED** [review]
Leave `Big` buggy here, I'll fix it in #8252.

**Comment by ejgallego** [comment]
> Leave `Big` buggy here, I'll fix it in #8252.

Indeed you can rebase #8252 now on top of this and things should fare better, as now `zarith` will be installed in the proper CI images.

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml:87]:
```suggestion
```

Comment (ejgallego) [reply interp/numTok.ml]: Thanks again @proux01 !

**Comment by maximedenes** [comment]
@ejgallego maybe you need to rebase to get rid of the fiat_parsers failure?

**Comment by ejgallego** [comment]
Rebased, before rebase CI was green so this is getting ready.

**Comment by ejgallego** [comment]
Failure on performance tests is spurious (c.f. #12615 )

**Review by SkySkimmer — APPROVED** [review]
WRT the build if it works then it should be OK

File: configure.ml
Comment (SkySkimmer) [inline configure.ml:699]:
```suggestion
  let zarith,_ = tryrun camlexec.find ["query";"zarith"] in
  match zarith with
```

Comment (ejgallego) [reply configure.ml]: Done thanks.

**Comment by maximedenes** [comment]
@ejgallego can you open a PR for the overlay?

**Comment by ejgallego** [comment]
> @ejgallego can you open a PR for the overlay?

Ups, done.

**Comment by liyishuai** [comment]
What's the status of this PR?

**Comment by ejgallego** [comment]
> @ejgallego would you mind rebasing?

Done.

**Comment by maximedenes** [comment]
@coqbot merge now

**Comment by maximedenes** [comment]
@coqbot Yes @coqbot, I shall do so. I hope you're having a good day. Thanks for merging!

**Comment by RalfJung** [comment]
Is it possible that this broke opam-based Coq builds? https://github.com/coq/coq/issues/12937

When dependencies change, would be good if the opam package was either fixed at the same time, or the opam maintainers were notified ahead of time.

File: INSTALL.md
Comment (RalfJung) [inline INSTALL.md:15]: The version given here is inconsistent with what it says in the opam file (1.9.1).

**Comment by bschommer** [comment]
> * windows build needs to install gmp + zarith

For windows, this [branch](https://github.com/ocaml/Zarith/tree/features/tommath-support) might be interesting since the libtommath builds fine under windows.

**Comment by MSoegtropIMC** [comment]
> > * windows build needs to install gmp + zarith
> 
> For windows, this [branch](https://github.com/ocaml/Zarith/tree/features/tommath-support) might be interesting since the libtommath builds fine under windows.

Not sure what the issue is on Windows. The Coq platform does build gmp on Windows. The opam package works fine.

**Comment by bschommer** [comment]
> > > * windows build needs to install gmp + zarith
> > 
> > 
> > For windows, this [branch](https://github.com/ocaml/Zarith/tree/features/tommath-support) might be interesting since the libtommath builds fine under windows.
> 
> Not sure what the issue is on Windows. The Coq platform does build gmp on Windows. The opam package works fine.

If one wants to use the msvc, one does not get to far with the gmp, mpir (the gmp fork with different license and build script for visual studio) gave up on releasing new versions. Of course gmp should normally work if you use mingw or cygwin.

**Comment by ejgallego** [comment]
@bschommer vendoring `zarith` would be an option, I'd be happy to finish https://github.com/ocaml/Zarith/pull/73 if the `zartih` devs give the OK.

**Pull request merged**
